tests/test_url.py

```diff
@@ -553,6 +553,86 @@ def test_canonicalize_url_idna_exceptions(self):
                     label=u"example"*11))
 
 
+class DataURITests(unittest.TestCase):
+
+    def test_default_mediatype_charset(self):
+        result = parse_data_uri("data:,A%20brief%20note")
+        self.assertEqual(result.media_type, "text/plain")
+        self.assertEqual(result.media_type_parameters, {"charset": "US-ASCII"})
+        self.assertEqual(result.data, b"A brief note")
+
+    def test_text_uri(self):
+        result = parse_data_uri(u"data:,A%20brief%20note")
+        self.assertEqual(result.data, b"A brief note")
+
+    def test_bytes_uri(self):
+        result = parse_data_uri(b"data:,A%20brief%20note")
+        self.assertEqual(result.data, b"A brief note")
+
+    def test_unicode_uri(self):
+        result = parse_data_uri(u"data:,é")
+        self.assertEqual(result.data, u"é".encode('utf-8'))
+
+    def test_default_mediatype(self):
+        result = parse_data_uri("data:;charset=iso-8859-7,%be%d3%be")
+        self.assertEqual(result.media_type, "text/plain")
+        self.assertEqual(result.media_type_parameters,
+                         {"charset": "iso-8859-7"})
+        self.assertEqual(result.data, b"\xbe\xd3\xbe")
+
+    def test_text_charset(self):
+        result = parse_data_uri("data:text/plain;charset=iso-8859-7,%be%d3%be")
+        self.assertEqual(result.media_type, "text/plain")
+        self.assertEqual(result.media_type_parameters,
+                         {"charset": "iso-8859-7"})
+        self.assertEqual(result.data, b"\xbe\xd3\xbe")
+
+    def test_mediatype_parameters(self):
+        result = parse_data_uri('data:text/plain;'
+                                'foo=%22foo;bar%5C%22%22;'
+                                'charset=utf-8;'
+                                'bar=%22foo;%5C%22foo%20;/%20,%22,'
+                                '%CE%8E%CE%A3%CE%8E')
+
+        self.assertEqual(result.media_type, "text/plain")
+        self.assertEqual(result.media_type_parameters,
+                         {"charset": "utf-8",
+                          "foo": 'foo;bar"',
+                          "bar": 'foo;"foo ;/ ,'})
+        self.assertEqual(result.data, b"\xce\x8e\xce\xa3\xce\x8e")
+
+    def test_base64(self):
+        result = parse_data_uri("data:text/plain;base64,"
+                                "SGVsbG8sIHdvcmxkLg%3D%3D")
+        self.assertEqual(result.media_type, "text/plain")
+        self.assertEqual(result.data, b"Hello, world.")
+
+    def test_base64_spaces(self):
+        result = parse_data_uri("data:text/plain;base64,SGVsb%20G8sIH%0A%20%20"
+                                "dvcm%20%20%20xk%20Lg%3D%0A%3D")
+        self.assertEqual(result.media_type, "text/plain")
+        self.assertEqual(result.data, b"Hello, world.")
+
+        result = parse_data_uri("data:text/plain;base64,SGVsb G8sIH\n  "
```

Comment:
imho, the implementation should support whitespace in base64 encoded data part.
I don't think we need to support whitespace elsewhere. (so that would mean following what Chrome does, I believe)